### DISCUSSION TITLE
Reason for the interfaces?

### DISCUSSION THREAD

**Discussion by damccull** [ravendb/ravendb#14249] [Q&A] [answered]
Hello,

Been wondering why the library has a bunch of interfaces if there only seems to be one concrete type for things like the DocumentStore...

Is this client meant to be user extensible? Or is this just done for easier test mocking? Some other reason?

Thanks.

**Comment by georgiosd** [comment]
Not sure if this still true since I haven't used the embeddable store recently, but that used to be a different type.

**Comment by ayende** [comment, marked as answer]
Right now, there is a single implementation, but there used to be more (embedded and server).
It also make it easier to isolate features, such as things that you are expected to do for initialization and for usage.

Part of that is also for making testing easier for those people who want to mock the database access (which I wouldn't recommend).
